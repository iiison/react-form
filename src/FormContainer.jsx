import React, { Component } from 'react'
import PropTypes            from 'prop-types' // eslint-disable-line import/no-extraneous-dependencies

import validations          from './utils/validationRules'

/**
 * Select rule from 3 diffrent places(predefined rules, form level rules or input level rules)
 */
export function selectRule({ ruleValue, customFormRules, customRules }) {
  return validations[ruleValue] || customRules[ruleValue] || customFormRules[ruleValue]
}

/**
 * Validates if the the selected rule(returned from `selectRule`) is valid or not.
 * if rule is valid then tests the input field value agains the selected rule.
 * returns empty string if rule is valid and saticefies the condition.
 */
export function validateRule({ id, value, displayName, ruleValue, ruleArgs, validation }) {
  if (validation) {
    const stringifiedValue = value.toString()
    const isRuleSatisfied = ( ruleValue !== 'required' && !stringifiedValue )
      ? true
      : validation.rule.apply(null, ruleArgs).test(value.toString())
    let error = ''

    if (!isRuleSatisfied) {
      error = validation.formatter.apply(null, [displayName || id, ...ruleArgs])

      return error
    }
  } else {
    throw `invalid validation rule: ${ruleValue}, please use an existing validation rule name or pass a custom function with same name through 'customRules' prop in Input: ${id}. Rule value should be an object with keys: 'rule' as an Regex and 'formatter' as a function, that formats the value.` // eslint-disable-line
  }

  return ''
}


/**
 * Main Component, a HOC, will store all fields' state.
 */
export default class FormContainer extends Component {
  constructor(props) {
    super(props)

    const {
      isDisabled = false,
      shouldValidateForm,
      defaultClasses
    } = this.props

    this.state = {
      shouldValidateForm,
      defaultClasses,
      isDisabled,
      errors     : {},
      isFetching : false
    }
  }

  getChildContext() {
    return {
      setFieldValue : ({ event, field, value, isMultipleValues, id = 'value' }, callBack) => {
        const fieldName = field.id
        const setState = (name, fieldValue) => {
          this.setState((prevState) => ({
            ...prevState,
            fields : {
              ...prevState.fields,
              [fieldName] : {
                ...prevState.fields[name],
                shouldValidateField : true,
                [id]                : fieldValue
              }
            }
          }), callBack)
        }

        if (!value && value !== '') {
          setState(fieldName, event.currentTarget.value)
        } else if (isMultipleValues) {
          const stateCopy = { ...this.state }

          for (const item in value) {
            stateCopy.fields[item].shouldValidateField = true
            stateCopy.fields[item][id] = value[item]
          }

          this.setState((prevState) => ({
            ...prevState,
            ...stateCopy
          }))
        } else {
          setState(fieldName, value)
        }

        return this.state
      },

      addField : (data) => {
        this.setState((prevState) => ({
          ...prevState,
          errors : {
            ...prevState.errors,
            [data.id] : ''
          },
          fields : {
            ...prevState.fields,
            [data.id] : {
              ...data,
              // shouldValidateField : false // Not sure why this was added
            }
          }
        }))
      },

      setFormData : (data) => {
        this.setState((prevState) => ({
          ...prevState,
          ...data
        }))
      },

      validateForm : (fieldName) => {
        const { shouldValidateForm, fields } = this.state
        const field = fields[fieldName]

        if (shouldValidateForm) {
          if (field) {
            if ( field.shouldValidateField ) {
              this.validateField(field)
            }

            return
          }

          for (const unit in fields) {
            const fieldData = fields[unit]

            if (fieldData) {
              this.validateField(fieldData)
            }
          }
        }

        return
      },
      formData : this.state
    }
  }

  /**
   * React Lifecycle Method: Renders the data
   *
   * @return {DOM} Main container DOM.
   */
  render() {
    const { children } = this.props

    return (
      <form action=''>
        {children}
      </form>
    )
  }

  validateField = ({
    id,
    value,
    validate,
    displayName,
    customRules = {}
  }) => {
    let error = ''
    const rules = validate ? validate.split('|') : ''
    const { customValidators : customFormRules } = this.props

    if (rules.length) {
      for (const rule in rules) {
        const ruleName = rules[rule]
        const ruleDetails = ruleName.split('-')
        const [ruleValue, ...ruleArgs] = ruleDetails
        const validation = selectRule({ ruleValue, customFormRules, customRules })

        try {
          error = validateRule({ id, value, displayName, ruleValue, ruleArgs, validation })

          if (error !== '') {
            break
          }
        } catch (invalidRuleError) {
          throw invalidRuleError
        } 
      }

      this.setState((prevState) => ({
        ...prevState,
        errors : {
          ...prevState.errors,
          [id] : error
        }
      }))
    }
  }

  static defaultProps = {
    errors             : {}, // eslint-disable-line
    children           : <div />,
    isDisabled         : false,
    shouldValidateForm : true,
    customValidators   : {},
    defaultClasses     : {
      contClass  : '',
      fieldClass : '',
      errorClass : '',
      labelClass : ''
    }
  }

  static propTypes = {
    children           : PropTypes.node,
    isDisabled         : PropTypes.bool,
    shouldValidateForm : PropTypes.bool,
    customValidators   : PropTypes.object, // eslint-disable-line
    defaultClasses     : PropTypes.shape({
      labelClass : PropTypes.string,
      contClass  : PropTypes.string,
      errorClass : PropTypes.string,
      fieldClass : PropTypes.string,
    })
  }

  static childContextTypes = {
    addField      : PropTypes.func,
    setFieldValue : PropTypes.func,
    validateForm  : PropTypes.func,
    formData      : PropTypes.object, // eslint-disable-line
    setFormData   : PropTypes.func
  }
}

