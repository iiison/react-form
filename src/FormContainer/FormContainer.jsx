import React, { Component } from 'react'

import { Form } from '$COMPONENTS'
import validations from '$UTILS/validationRules'

/**
 * Test all validation rules ofa field
 *
 * @param {Object} fieldData all field realated data
 *
 * @return {String}         validation error ina field
 */
export function testFieldRules(fieldData) {
  const { validate, value, displayName } = fieldData

  const validationRules = validate.split('|')

  for (const validation in validationRules) {
    const { rule, formatter } = validations[validationRules[validation]]()

    const result = rule.test(value)

    if (result === false) {
      return formatter(displayName)
    }
  }
}

/**
 * Generic Form Handler, Handles form state,
 * Pass Form elemnt's JSON throught the FormContainer and
 * it will redner the form.
 */
export default class FormContainer extends Component {
  /**
   * Constructor for the container
   *
   * @param {Object} props form data
   */
  constructor(props) {
    super(props)
    const { formData : { fields } } = props

    this.state = {
      isTested : false,
    }

    this.state.fields = fields.reduce((prev, curr) => {
      const temp = {
        value          : '',
        error          : '',
        shouldBeTested : false,
        ...curr
      }

      return {
        ...prev,
        [temp.id] : temp
      }
    }, {})
  }

  /**
   * DOM Event : Chaanges `shouldBeTested`, will be executed on
   * keyUp in an input field
   *
   * @param {String} field field name
   */
  makeFieldTestable = (field) => {
    const fieldRef = this.state.fields[field]

    fieldRef.shouldBeTested = true

    this.setState((previousState) => {
      return {
        fields : {
          ...previousState.fields,
          [field] : fieldRef
        }
      }
    })
  }

  /**
   * Handle Key press event on the input,
   * tests field value if field is already has error.
   * @param  {String} field Field name in which event was started
   */
  handleKeyPress = (field) => {
    const fieldData = this.state.fields[field]
    const { error } = fieldData

    this.makeFieldTestable(field)

    if (error) {
      this.validateInput(field)
    }
  }

  /**
   * DOM Event: Handles input value change
   *
   * @param {Object} event DOM event object
   */
  handleChange = (event) => {
    event.preventDefault()

    const elemRef = event.target
    const idName = elemRef.id

    this.setState((previousState) => {
      const { fields } = previousState

      return {
        fields : {
          ...fields,
          [idName] : {
            ...fields[idName],
            value : elemRef.value
          }
        }
      }
    })
  }

  /**
   * Validate single input
   *
   * @param  {String}  field       field name to be tested
   * @param  {Boolean} isFinalTest will be true if whole form is being tested
   * @return {String}              validation error of the field
   */
  validateInput = (field, isFinalTest) => {
    const fieldData = this.state.fields[field]
    const { validate } = fieldData

    if ((fieldData.shouldBeTested === true || isFinalTest) && validate) {
      const error = testFieldRules(fieldData) || ''

      this.setState((previousState) => {
        const { fields } = previousState
        const updatedField = {
          ...fields[field],
          error
        }

        return {
          fields : {
            ...fields,
            [field] : updatedField
          }
        }
      })

      return error
    }
  }

  /**
   * Handles Form Submission, callback actual form submit fxn
   * after the form validation. validation is the part of plugin.
   *
   * @param {Object} event DOM Event object
   */
  handleFormSubmit = (event) => {
    event.preventDefault()

    const { handleSubmit } = this.props
    const { fields } = this.state
    const errors = []

    for ( const field in fields ) {
      const error = this.validateInput(field, true)

      error && errors.push(error)
    }

    if (!errors.length) {
      handleSubmit(fields)
    }
  }

  /**
   * React Lifecycle Method: Renders the data
   *
   * @return {DOM} Main container DOM.
   */
  render() {
    const { fields } = this.state
    const {
      error,
      messages,
      isFetching,
      submitButtonText,
      submitButtonClassName } = this.props

    const props = {
      fields,
      messages,
      events : {
        onSubmit          : this.handleFormSubmit,
        onValueChange     : this.handleChange,
        validateInput     : this.validateInput,
        handleKeyPress    : this.handleKeyPress,
        makeFieldTestable : this.makeFieldTestable
      },
      submitButtonData : {
        error,
        isFetching,
        submitButtonText,
        submitButtonClassName,
        onSubmit : this.handleFormSubmit
      }
    }

    return (
      <Form
        {...props}
      />
    )
  }

  /**
  * React Lifecycle Method: Executes right before the
  * component mounts
  */
  componentWillMount() {
  }
}
