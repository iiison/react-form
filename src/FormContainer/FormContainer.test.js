import React               from 'react'
import renderer            from 'react-test-renderer'
import { mount, shallow }  from 'enzyme'

import loginForm           from '$CONFIG/forms/loginForm.json'
import FormContainer,
{ testFieldRules }         from './FormContainer'

// import loginMessages               from './lang/en.json'

const props = {
  error                 : '',
  isFetching            : false,
  formData              : loginForm,
  handleSubmit          : jest.fn(),
  submitButtonText      : 'Test',
  submitButtonClassName : `button`
}

/* eslint-disable react/jsx-filename-extension */

// Snapshot matching for Login Container
describe('>>> FormContainer Container -- Snapshot Test', () => {
  const initialState = props

  it('Matches the snapshot with isFetching is false', () => {
    const tree = renderer.create(<FormContainer {...initialState} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot when isFetching is true', () => {
    const updatedState = {
      ...props,
      isFetching : true
    }

    const tree = renderer.create(<FormContainer {...updatedState} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot when error occured', () => {
    const updatedState = {
      ...props,
      error : 'not working'
    }

    const tree = renderer.create(<FormContainer {...updatedState} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
// *************************************************************

/* eslint-enable */


// Check DOM Behaviour
describe('>>> Form Container -- Shallow Rendering', () => {
  const initialState = props
  let mountedScreen
  const FormScreen = (customState = {}) => {
    if (!mountedScreen) {
      const updatedState = {
        ...initialState,
        ...customState
      }

      mountedScreen = mount(<FormContainer {...updatedState} />)
    }

    return mountedScreen
  }

  beforeEach(() => {
    mountedScreen = undefined
  })

  it('Test and confirm Button Text.', () => {
    const domTree = FormScreen()

    const expectedButtonText = 'test'
    const buttonText = domTree
      .find('button')
      .first()
      .text()
      .toLowerCase()

    expect(buttonText).toEqual(expectedButtonText)
  })

  it('Test and confirm Button Text when isFetching is true.', () => {
    const domTree = FormScreen({
      isFetching : true
    })
    const expectedButtonText = 'loading...'
    const buttonText = domTree
      .find('button')
      .first()
      .text()
      .toLowerCase()

    expect(buttonText).toEqual(expectedButtonText)
  })

  it('Creates form submission error.', () => {
    const error = 'some random error'
    const domTree = FormScreen({ error })
    const expectedButtonText = 'loading...'

    const errorText = domTree
      .find('p.error')
      .first()
      .text()

    expect(errorText).toEqual(error)
  })

  it('Tests and confirms input count and text.', () => {
    const domTree = FormScreen()
    const expectedInputLength = 2
    const inputs = domTree
      .find('input')

    const firstInputProps = inputs.at(0).props()
    const secondInputProps = inputs.at(1).props()

    let expectedType = 'email'
    expect(firstInputProps.type).toEqual(expectedType)

    expectedType = 'password'
    expect(secondInputProps.type).toEqual(expectedType)

    let expectedPlaceholder = ' '
    expect(firstInputProps.placeholder).toEqual(expectedPlaceholder)

    expectedPlaceholder = ' '
    expect(secondInputProps.placeholder).toEqual(expectedPlaceholder)

    const placeholders = domTree.find('.placeholder')

    const firstPlaceholderText = placeholders.at(0).text()
    const secondPlaceholderText = placeholders.at(1).text()

    let expectedPleaceholderText = 'enter email'
    expect(firstPlaceholderText).toEqual(expectedPleaceholderText)

    expectedPleaceholderText = 'enter password'
    expect(secondPlaceholderText).toEqual(expectedPleaceholderText)
  })

  it('Simulates text change event in an input', () => {
    const getDashboardItems = jest.fn()
    const tree = FormScreen()

    const instance = tree.instance()
    const handleInputChangeSpy = jest.spyOn(instance, 'handleChange')

    const input = tree
      .find('input')
      .first()

    input.instance().value = 'someValue'
    instance.forceUpdate()

    input.simulate('change')

    expect(handleInputChangeSpy).toHaveBeenCalled()
  })

  it('Simulates text key press event in an input', () => {
    const getDashboardItems = jest.fn()
    const tree = FormScreen()

    const instance = tree.instance()
    const handleKeyPressSpy = jest.spyOn(instance, 'handleKeyPress')

    const input = tree
      .find('input')
      .first()

    instance.forceUpdate()
    input.simulate('keyPress')
    expect(handleKeyPressSpy).toHaveBeenCalled()
  })

  it('Simulates submit button click', () => {
    const getDashboardItems = jest.fn()
    const tree = FormScreen()

    const instance = tree.instance()
    const handleFormSubmitSpy = jest.spyOn(instance, 'handleFormSubmit')
    const validateInputSpy = jest.spyOn(instance, 'validateInput')

    const input = tree
      .find('button')
      .first()

    instance.forceUpdate()
    input.simulate('click')

    expect(handleFormSubmitSpy).toHaveBeenCalled()
    expect(validateInputSpy.mock.calls).toEqual([['userName', true], ['password', true]])
  })

  it('Tests email Validation Rules for a field with valid input', () => {
    const result = testFieldRules({
      validate    : 'email',
      value       : 'test@gmail.com',
      displayName : 'Test'
    })

    expect(result).toBeUndefined()
  })

  it('Tests email Validation Rules for a field with invalid input', () => {
    const result = testFieldRules({
      validate    : 'email',
      value       : 'test@om',
      displayName : 'Test'
    })
    const expectedResult = 'Test is not valid'

    expect(result).toEqual(expectedResult)
  })

  it('Tests required Validation Rule for a field with valid input', () => {
    const result = testFieldRules({
      validate    : 'required',
      value       : 'some value',
      displayName : 'Test'
    })

    expect(result).toBeUndefined()
  })

  it('Tests required Validation Rule for a field with invalid input', () => {
    const result = testFieldRules({
      validate    : 'required',
      value       : '',
      displayName : 'Test'
    })
    const expectedResult = 'Test is required'

    expect(result).toEqual(expectedResult)
  })
})
// *************************************************************
