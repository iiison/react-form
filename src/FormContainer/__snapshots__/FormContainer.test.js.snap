// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> FormContainer Container -- Snapshot Test Matches the snapshot when error occured 1`] = `
<form
  action=""
  className=""
>
  <div
    className="input-container"
  >
    <div
      className="textContainer"
      data-placeholder="enter email"
    >
      <input
        className="text  undefined"
        id="userName"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder=" "
        type="email"
        value=""
      />
      <span
        className="placeholder"
      >
        enter email
      </span>
    </div>
    <span
      className="error"
    >
      
    </span>
  </div>
  <div
    className="input-container"
  >
    <div
      className="textContainer"
      data-placeholder="enter password"
    >
      <input
        className="text  undefined"
        id="password"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder=" "
        type="password"
        value=""
      />
      <span
        className="placeholder"
      >
        enter password
      </span>
    </div>
    <span
      className="error"
    >
      
    </span>
  </div>
  <div
    className="col"
  >
    <button
      className="button"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Test
    </button>
    <p
      className="error"
    >
      not working
    </p>
  </div>
</form>
`;

exports[`>>> FormContainer Container -- Snapshot Test Matches the snapshot when isFetching is true 1`] = `
<form
  action=""
  className=""
>
  <div
    className="input-container"
  >
    <div
      className="textContainer"
      data-placeholder="enter email"
    >
      <input
        className="text  undefined"
        id="userName"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder=" "
        type="email"
        value=""
      />
      <span
        className="placeholder"
      >
        enter email
      </span>
    </div>
    <span
      className="error"
    >
      
    </span>
  </div>
  <div
    className="input-container"
  >
    <div
      className="textContainer"
      data-placeholder="enter password"
    >
      <input
        className="text  undefined"
        id="password"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder=" "
        type="password"
        value=""
      />
      <span
        className="placeholder"
      >
        enter password
      </span>
    </div>
    <span
      className="error"
    >
      
    </span>
  </div>
  <div
    className="col"
  >
    <button
      className="button"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      loading...
    </button>
    <p
      className="error"
    >
      
    </p>
  </div>
</form>
`;

exports[`>>> FormContainer Container -- Snapshot Test Matches the snapshot with isFetching is false 1`] = `
<form
  action=""
  className=""
>
  <div
    className="input-container"
  >
    <div
      className="textContainer"
      data-placeholder="enter email"
    >
      <input
        className="text  undefined"
        id="userName"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder=" "
        type="email"
        value=""
      />
      <span
        className="placeholder"
      >
        enter email
      </span>
    </div>
    <span
      className="error"
    >
      
    </span>
  </div>
  <div
    className="input-container"
  >
    <div
      className="textContainer"
      data-placeholder="enter password"
    >
      <input
        className="text  undefined"
        id="password"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder=" "
        type="password"
        value=""
      />
      <span
        className="placeholder"
      >
        enter password
      </span>
    </div>
    <span
      className="error"
    >
      
    </span>
  </div>
  <div
    className="col"
  >
    <button
      className="button"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Test
    </button>
    <p
      className="error"
    >
      
    </p>
  </div>
</form>
`;
